doc/developers/cython.rst

```diff
@@ -0,0 +1,152 @@
+.. _cython:
+
+Cython Best Practices, Conventions and Knowledge
+================================================
+
+This documents:
+
+* technical aspects of Cython which aren't currently present in Cython's documentation
+* tips to develop Cython code in scikit-learn
+
+Tips for developing with Cython in scikit-learn
+-----------------------------------------------
+
+Tips to ease development
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+* Time spent reading `Cython's documentation <https://cython.readthedocs.io/en/latest/>`_ is not time lost.
+
+* If you intend to use OpenMP: On MacOS, system's distribution of ``clang`` does not implement OpenMP.
+  You can install the ``compilers`` package available on ``conda-forge`` which comes with implementations of OpenMP.
+
+* Activating `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_ might help. E.g. for activating boundscheck use:
+
+  .. code-block:: bash
+
+         export SKLEARN_ENABLE_DEBUG_CYTHON_DIRECTIVES=1
+
+* `Start from scratch in a notebook <https://cython.readthedocs.io/en/latest/src/quickstart/build.html#using-the-jupyter-notebook>`_ to understand how to use Cython and to get feedback on your work quickly.
+
+* To debug C code (e.g. a segfault), do use ``gdb`` with:
+
+  .. code-block:: bash
+
+         gdb --ex r --args python ./entrypoint_to_bug_reproducer.py
+
+* To have access to some value in place to debug in ``cdef (nogil)`` context, use:
+
+  .. code-block:: cython
+
+         with gil:
+             print(state_to_print)
+
+* Note that Cython cannot parse f-strings with ``{var=}`` expressions, e.g.
+
+  .. code-block:: bash
+
+         print(f"{test_val=}")
+
+* scikit-learn codebase has a lot of non-unified (fused) types (re)definitions. There currently is `ongoing work to simplify
+  and unify that across the codebase <https://github.com/scikit-learn/scikit-learn/issues/25572>`_. For now, make sure you understand which concrete types are used ultimately.
+
+* It is helpful to use ``gdb`` to debug. In order to do so, one must use a Python interpreter built with debug support
+  (debug symbols and proper optimization). To create a new conda environment (which you might need to deactivate and
+  reactivate conda after building/installing) with a source-built CPython interpreter:
```

Comment:
I'm still not sure that this section belongs to cython tips

Reply:
I'll move it to another file.